Update thumbnail loading assets

rgozim opened the pull request:
# What this PR does

Adds new loading and error thumbnail assets to Inisght

rgozim committed:
Added loading and error asset in place of rendered text

rgozim committed:
Using createScaledImage to check for concurrent import with thumbnail loading

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/env/data/views/calls/ThumbnailLoader.java:
```diff
@@ -241,24 +226,15 @@ public void doCall() throws Exception {
-                            currentThumbnail = new ThumbnailData(pxd.getImage()
-                                    .getId(), thumbnail, userId, true);
+                        } catch (DSAccessException e) {
+                            currentThumbnail = Factory.createThumbnailFromRes("icons/error_black_96px.png");
```
Icon should be retrieved from the icon manager
org.openmicroscopy.shoola.env.ui.IconManager

jburel replied:
sorry org.openmicroscopy.shoola.util.ui I copy the wrong link

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/util/image/geom/Factory.java:
```diff
@@ -284,6 +286,22 @@ public static BufferedImage createDefaultThumbnail(String text)
+	 * @return A BufferedImage if resource found, otherwise null
+	 * @throws IOException if an error occurs during reading
+	 */
+	public static BufferedImage createThumbnailFromRes(String resUrl) throws IOException {
```
a more explicit name will be useful
i.e. ``createThumbnailFromResource``

snoopycrimecop commented:
Conflicting PR. Removed from build [OMERO-DEV-merge-push#1106](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/1106/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/1106/consoleText) for more details.
Possible conflicts:
  - PR #5852 dominikl 'Lookupnames'
    - components/insight/SRC/org/openmicroscopy/shoola/env/data/views/calls/ThumbnailLoader.java

~~--conflicts~~ Conflict resolved in build [OMERO-DEV-merge-push#1107](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/1107/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/1107/consoleText) for more details.

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/env/data/views/calls/ThumbnailLoader.java:
```diff
@@ -34,16 +35,19 @@
 import omero.gateway.model.PixelsData;
 import omero.log.LogMessage;
 
+
```
don't need this extra blank

mtbc commented:
Warning: not in today's West build.

mtbc commented:
From eel merge with OMERO.insight-5.4.8-rc2-35-dccfb48-ice35-b1075 I tried looking at a big image I'd imported and got,
```
java.lang.Exception: Abnormal termination due to an uncaught exception.
java.lang.NullPointerException: No thumbnail.
	at org.openmicroscopy.shoola.agents.dataBrowser.ThumbnailsManager.setThumbnail(ThumbnailsManager.java:153)
	at org.openmicroscopy.shoola.agents.dataBrowser.view.DataBrowserModel.setThumbnail(DataBrowserModel.java:502)
	at org.openmicroscopy.shoola.agents.dataBrowser.view.DataBrowserComponent.setThumbnail(DataBrowserComponent.java:287)
	at org.openmicroscopy.shoola.agents.dataBrowser.ThumbnailLoader.update(ThumbnailLoader.java:212)
	at org.openmicroscopy.shoola.env.data.events.DSCallAdapter.eventFired(DSCallAdapter.java:75)
	at org.openmicroscopy.shoola.env.data.views.BatchCallMonitor$1.run(BatchCallMonitor.java:124)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
...
```

mtbc commented:
Gah, sorry, that looks like the wrong build to me, let me redownload and check. *Update:* It was the right build, I just looked at the wrong part of my terminal scrollback!

rgozim committed:
Numerous changes
Removed Factory.createThumbnailFromRes
Moved icon assets from resources folder to graphx (match project style)
Used IconManager in ThumbnailLoader to load icon assets

joshmoore commented on components/insight/SRC/org/openmicroscopy/shoola/env/data/views/calls/ThumbnailLoader.java:
```diff
@@ -235,24 +223,15 @@ public void doCall() throws Exception {
-                                    .createDefaultThumbnail("Can't render\n[Error "+Error.NO_THUMBNAILSTORE.code+"]");
-                            currentThumbnail = new ThumbnailData(pxd.getImage()
-                                    .getId(), thumbnail, userId, true);
+                        } catch (DSAccessException e) {
```
Why the change in the exception handling? Throwable was specifically added in https://github.com/openmicroscopy/openmicroscopy/pull/5846/files/0d3ef709a783913874239a39d186c4b1d9f45596#r215894053

mtbc commented:
https://github.com/openmicroscopy/openmicroscopy/pull/5854#issuecomment-420566456 is now fixed and I see a "loading" thumbnail:
![loading](https://user-images.githubusercontent.com/2630707/45418734-230caa80-b67d-11e8-91a5-3b38ad27d3aa.png)

mtbc commented:
Looking at a thumbnail-less image, for which I do not have permission to create my own, Insight crashes with:
```
java.lang.Exception: omero.ApiUsageException
    serverStackTrace = "ome.conditions.ApiUsageException: Unable to reset rendering settings in a read-only group for Pixels set id:149762
                        	at ome.services.ThumbnailBean.resetDefaults(ThumbnailBean.java:1443)
                        	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
                        	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
                        	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                        	at java.lang.reflect.Method.invoke(Method.java:606)
                        	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
                        	at ome.security.basic.EventHandler.invoke(EventHandler.java:154)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.tools.hibernate.SessionHandler.doStateful(SessionHandler.java:216)
                        	at ome.tools.hibernate.SessionHandler.invoke(SessionHandler.java:200)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:108)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.tools.hibernate.ProxyCleanupFilter$Interceptor.invoke(ProxyCleanupFilter.java:249)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.services.util.ServiceHandler.invoke(ServiceHandler.java:121)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
                        	at com.sun.proxy.$Proxy102.resetDefaults(Unknown Source)
                        	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
                        	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
                        	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                        	at java.lang.reflect.Method.invoke(Method.java:606)
                        	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
                        	at ome.security.basic.BasicSecurityWiring.invoke(BasicSecurityWiring.java:93)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at ome.services.blitz.fire.AopContextInitializer.invoke(AopContextInitializer.java:43)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
                        	at com.sun.proxy.$Proxy102.resetDefaults(Unknown Source)
                        	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
                        	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
                        	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                        	at java.lang.reflect.Method.invoke(Method.java:606)
                        	at ome.services.blitz.util.IceMethodInvoker.invoke(IceMethodInvoker.java:172)
                        	at ome.services.throttling.Callback.run(Callback.java:56)
                        	at ome.services.throttling.InThreadThrottlingStrategy.callInvokerOnRawArgs(InThreadThrottlingStrategy.java:56)
                        	at ome.services.blitz.impl.AbstractAmdServant.callInvokerOnRawArgs(AbstractAmdServant.java:140)
                        	at ome.services.blitz.impl.ThumbnailStoreI.resetDefaults_async(ThumbnailStoreI.java:138)
                        	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
                        	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
                        	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
                        	at java.lang.reflect.Method.invoke(Method.java:606)
                        	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:307)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:183)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:150)
                        	at omero.cmd.CallContext.invoke(CallContext.java:85)
                        	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
                        	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
                        	at com.sun.proxy.$Proxy103.resetDefaults_async(Unknown Source)
                        	at omero.api._ThumbnailStoreTie.resetDefaults_async(_ThumbnailStoreTie.java:174)
                        	at omero.api._ThumbnailStoreDisp.___resetDefaults(_ThumbnailStoreDisp.java:843)
                        	at omero.api._ThumbnailStoreDisp.__dispatch(_ThumbnailStoreDisp.java:982)
                        	at IceInternal.Incoming.invoke(Incoming.java:222)
                        	at Ice.ConnectionI.invokeAll(ConnectionI.java:2482)
                        	at Ice.ConnectionI.dispatch(ConnectionI.java:1258)
                        	at Ice.ConnectionI.message(ConnectionI.java:1213)
                        	at IceInternal.ThreadPool.run(ThreadPool.java:321)
                        	at IceInternal.ThreadPool.access$300(ThreadPool.java:12)
                        	at IceInternal.ThreadPool$EventHandlerThread.run(ThreadPool.java:693)
                        	at java.lang.Thread.run(Thread.java:745)
                        "
...
```

rgozim committed:
Added catch for ServerError
Removed unused code

mtbc commented:
For that same case I'm now getting,
```
java.lang.ClassCastException: sun.awt.image.ToolkitImage cannot be cast to org.openmicroscopy.shoola.env.data.model.ThumbnailData
	at org.openmicroscopy.shoola.agents.dataBrowser.ThumbnailLoader.update(ThumbnailLoader.java:208)
	at org.openmicroscopy.shoola.env.data.events.DSCallAdapter.eventFired(DSCallAdapter.java:75)
	at org.openmicroscopy.shoola.env.data.views.BatchCallMonitor$1.run(BatchCallMonitor.java:124)
	at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at org.GNOME.Accessibility.AtkWrapper$5.dispatchEvent(AtkWrapper.java:700)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

dominikl commented on components/insight/SRC/org/openmicroscopy/shoola/env/data/views/calls/ThumbnailLoader.java:
```diff
@@ -235,24 +218,15 @@ public void doCall() throws Exception {
-                            currentThumbnail = new ThumbnailData(pxd.getImage()
-                                    .getId(), thumbnail, userId, true);
+                        } catch (DSAccessException | ServerError e) {
+                            currentThumbnail = getErrorIcon();
```
That's the issue for the ClassCastException noticed by @mtbc . Although an 'Object' currentThumbnail is assumed to be a 'ThumbnailData'.

dominikl replied:
`currentThumbnail = new ThumbnailData(pxd.getImage().getId(), getErrorIcon(), userId, false);` That should work.

rgozim replied:
Well spotted!

rgozim committed:
Removed whitespace, unused import and fixed type error in catch block

mtbc commented:
No stack trace any more, much nicer now, thanks:
![error](https://user-images.githubusercontent.com/2630707/45431247-2d3fa080-b69f-11e8-8320-2586cb861181.png)

jburel commented:
One question: where are the icons coming from?
If they are coming from a 3rd party location e.g. eclipse the name should be  ``source_nameoficon``
and declared in ``lib/licenses/README``

rgozim committed:
Updated README to include material.io reference and renamed assets
to start with material_

rgozim committed:
Updated IconManager

jburel commented on components/insight/SRC/org/openmicroscopy/shoola/util/ui/IconManager.java:
```diff
@@ -727,6 +733,8 @@
         relPaths[ROI_FOLDER_NOT_OWNER] = "roi_folder_user_icon.png";
         relPaths[LOOKUP_TABLE_24] = "lookuptable24.png";
         relPaths[LUTS] = "lookuptables.png";
+        relPaths[THUMBNAIL_ERROR_BLACK] = "thumbnail_error_black_96px.png";
```
the name needs to be changed here too

jburel commented:
tested the last commits locally
<img width="336" alt="pyramid_generation" src="https://user-images.githubusercontent.com/1022396/45480540-24071000-b741-11e8-812f-895ea71603c6.png">
<img width="530" alt="pyramid_error" src="https://user-images.githubusercontent.com/1022396/45480744-9bd53a80-b741-11e8-90d0-d422b5cbbed6.png">

insight changes look good

joshmoore commented:
Cheers.

jburel commented:
Some typo in the Java doc 
That can be fixed after release

jburel commented:
https://github.com/openmicroscopy/openmicroscopy/pull/5857
